MoodyCamel lockfree

biddisco opened the pull request:
This PR replaces some of our usage of boost::lockfree fifo queue with an alternative lockfree queue from MoodyCamel. Please see the commit log for extra info.

Improvements come in two places. 1) we get better performance with our task queues as this queue outperforms the boost one. Particularly on higher thread count numbers. 2) The removal of an alloc/dealloc in the task description queue reduces work in jemalloc/other allocators which can hit posix locks when they are maxxed out (particularly noticable in the future_overhead test for example).

I would like to have approval from @cameron314 that the inclusion of the boost license in the concurrentqueue.h file is in accordance with his terms and conditions.

msimberg commented on hpx/runtime/threads/policies/lockfree_queue_backends.hpp:
```diff
@@ -11,6 +11,7 @@
 #include <hpx/config.hpp>
 
 #include <hpx/util/lockfree/deque.hpp>
+#include <hpx/concurrent/concurrentqueue.h>
```
Why not `hpx/util/lockfree/concurrentqueue.h` like the deque?

biddisco replied:
When we were talking about integrating libcds with hpx, we had the idea of making hpx/concurrent as a new place for concurrent stuff. But I'm quite happy to put it into `util/lockfree/` since its a single header

msimberg replied:
Or move deque into concurrent?

biddisco replied:
Either works for me

sithhell replied:
I don't mind. The 'concurrent/concurrentqueue.h' has lots of redundant information. Also, please use the `.hpp` file ending to stay consinstent with the rest.

biddisco replied:
I've renamed the file to `.hpp` extension and moved it to hpx/util/lockfree/ - force pushed with the commits edited accordingly.

msimberg commented on hpx/runtime/threads/policies/thread_queue.hpp:
```diff
@@ -196,8 +196,8 @@ namespace hpx { namespace threads { namespace policies
 
-        typedef typename StagedQueuing::template
-            apply<task_description*>::type task_items_type;
+        typedef lockfree_fifo::
```
Either `StagedQueueing` should go from the template, or the other queues should still be supported?

biddisco replied:
StagedQueuing is still used for the thread_alloc_ template.
Not sure if it's actually needed though.

msimberg replied:
It's not actually used, so it could go.

biddisco replied:
I'll clean it up some more then.

Do @hkaiser or @sithhell want to comment on whether we should move lockfree stuff into concurrent, or just move the concurrentqueue into util/lockfree for now. (seems like putting it into util/lockfree for the time being is probably the simplest option, though I still like the idea of having a concurrent top level dir for this stuff)

hkaiser replied:
I'm impartial to this. Having it in util/lockfree is fine, I guess.

msimberg commented on hpx/concurrent/concurrentqueue.h:
```diff
@@ -0,0 +1,3648 @@
+                if (freeListHead.compare_exchange_strong(head, next, std::memory_order_acquire, std::memory_order_relaxed)) {
+                    // Yay, got the node. This means it was on the list, which means shouldBeOnFreeList must be false no
+                    // matter the refcount (because nobody else knows it's been taken off yet, it can't have been put back on).
+                    assert((head->freeListRefs.
```
If we don't want to support the other queue backends for terminated items, this one could be useful for `cleanup_terminated`.

biddisco replied:
I thought I'd start by changing the drop in replacments from fifo to new queue. The lifo stuff I left for now as I still have benchmarking I'm doing across schedulers that might help us decide on whether we want to remove other stuff.

msimberg replied:
Sure. I'm only specifically talking about `terminated_items_` since other queue backends can't be used there anymore.

biddisco replied:
ok. I did change `typename TerminatedQueuing = lockfree_fifo` in the shared priority scheduler to use it but I didn't realize that it was the only one we should use. the other schedulers are using lifo at the moment and I didn't touch them. I'll look again at the code.

hkaiser commented on hpx/runtime/threads/thread_init_data.hpp:
```diff
@@ -50,6 +50,31 @@ namespace hpx { namespace threads
+#endif
+            return *this;
+        }
+
```
Good catch!

hkaiser commented:
@biddisco Have you thought about using the advanced stuff in the ModyCamel queue (https://github.com/cameron314/concurrentqueue#advanced-features) to speed up stealing? We know what thread we will steal from, so we could use the producer's token. Not sure how much this would help, though.

biddisco commented:
> @biddisco Have you thought about using the advanced stuff in the ModyCamel queue (https://github.com/cameron314/concurrentqueue#advanced-features) to speed up stealing? We know what thread we will steal from, so we could use the producer's token. Not sure how much this would help, though.

Yes. I will look at using producer and consumer tokens. Adding them should be straightforward as we can assign them to each thread/queue at startup. I want to get a basic working version in first and once we are happy that nothing is broken, then make more improvements.

biddisco commented:
What we really need is a dequeue version of the queue so that we can support lifo/fifo operation and steal from the stale end and keep hot items on the same thread.

cameron314 commented:
Queue author here, license and comments inserted at top of concurrentqueue.h are fine :-)

biddisco commented:
> 
> 
> Queue author here, license and comments inserted at top of concurrentqueue.h are fine :-)

Thank you very much. I will also look though our contributors section and add a thanks to you in there somewhere. (assuming this PR gets merged)

Since I have your attention - would it be a lot of work to turn the queue into a dequeue so we can pull from either end? As noted in an earlier comment, when tasks are stolen from one thread queue to another, stealing from the cold end is preferred, but on the owning thread, stealing from the hot end can improve cache reuse. Is this something that might one day be supported?

cameron314 commented:
Sorry, turning the queue into a dequeue is pretty much impossible. It would need to be completely redesigned.

Do you need the fifo-ness of a queue? This queue provides fifo order only per-producer, with no global order. If you don't need order at all, the internal free-list data structure is probably closer to what you want, since it pops out hot items where possible.

biddisco commented:
> Do you need the fifo-ness of a queue? This queue provides fifo order only per-producer, with no global order. If you don't need order at all, the internal free-list data structure is probably closer to what you want, since it pops out hot items where possible.

Thanks for he information. The fact that the queue isn't linearizable might cause some unexpected behaviour in terms of task ordering, but it should not be a deal breaker for us since each core is generally maintaining its own queue and stealing takes place when the queues drain on one core. It is mostly the ability to steal from the cold end (of another thread's queue) rather than the hot one that is of interest, (but I don't have much in the way of numbers to back up my claim anyway).

sithhell commented:
I think non ordering of tasks in the queues is totally fine. After all, it looks like we have, and only require, weak consistency here, which is IMHO a good thing as it naturally reflects the actual scheduling of "stuff" in the hardware, I actually like that "hot" items are preferred. I think this is also where the speed comes from?

msimberg commented:
This looks good to me, but there are a lot of timeouts on CircleCI. It looks like you're up to date with master, so you have the latest bugfixes. It would be good to figure out if they're the fault of this PR (I don't spot anything odd), or if the new queue is triggering more race conditions because of changed timings.

I also want to make sure everyone is aware and on-board with hard-coding the staged queues to the new queue backend (if we want to keep support for the others I guess it would be doable with various helpers and wrappers). It looks like in this case the implementation matters more than any theoretical benefits we'd get from LIFO or ABP LIFO, but all we have so far is results from the future overhead test.

biddisco commented:
I'm fixing inspect failures for the PR and the clang-tidy reported problems, but I have a question.
Inspect does not like the many long lines in the new queue. I would like to make the queue exempt from our inspect checks because there is a strong possibility that we will want to check for new versions reasonably regularly and update the `concurrentqueue.hpp` file without having to make changes every time.
1) does anyone object to making the queue inspect exempt
2) if we are ok to do that, is there a file somewhere I can add the queue to to make it exempt (I think there is but I'm too lazy to look and am hoping someone can just tell me).

biddisco commented:
Force-pushed this PR again : I've fixed inspect errors, added `// NOLINT` to the offending clang-tidy errors in ConcurrentQueue and opened an issue on the original repo here https://github.com/cameron314/concurrentqueue/issues/149 just to make sure the clang-tidy problems are noted.

msimberg commented:
>     2. if we are ok to do that, is there a file somewhere I can add the queue to to make it exempt (I think there is but I'm too lazy to look and am hoping someone can just tell me).
I think adding `hpx-no-inspect` to the top of the file should do it: https://github.com/STEllAR-GROUP/hpx/blob/cde5fd7adfaae9570ad6d3139e8f8a2e8ac057fd/tools/inspect/inspect.cpp#L21-L22.

msimberg commented:
> Sorry, turning the queue into a dequeue is pretty much impossible. It would need to be completely redesigned.
@cameron314 sorry to bug you more, but what about LIFO (i.e. stack)?

cameron314 commented:
@msimberg Nope, same. Like most lock-free data structures, it's pretty difficult to turn into anything it's not already :-)

msimberg commented:
@biddisco started a couple of builds on rostam:
- http://rostam.cct.lsu.edu/builders/hpx_clang_6_boost_1_68_centos_x86_64_debug/builds/139
- http://rostam.cct.lsu.edu/builders/hpx_clang_6_boost_1_68_centos_x86_64_release/builds/143

Doesn't tell us what's wrong, but these tests seem to fail regularly:
- tests.unit.parallel.executors.thread_pool_executors
- tests.unit.parallel.executors.timed_thread_pool_executors
- tests.unit.threads.resource_manager
- tests.regressions.components_dir.distributed.tcp.bulk_new_3054

biddisco commented:
I have modified this PR to include the MoodyCamel concurrentqueue as an additional lockfree backend and not replace the existing lockfree fifo. Then switched the shared_priority scheduler aloone to use it in one place. This allows us to incrementally add support/use of this backend without changing other schedulers and breaking the failing executor tests that caused problems with the previous version of the PR

biddisco commented:
I've cleaned this up to fix all unexpected header test fails etc and it should be mergeable.
There are occasional dashboard fails, but they are most likely caused by other random problems with builds etc. This PR should not touch any existing tests unless they explicitly use the shared-priority scheduler.

hkaiser commented on hpx/runtime/threads/policies/lockfree_queue_backends.hpp:
```diff
@@ -17,8 +17,12 @@
 #include <boost/lockfree/queue.hpp>
 #endif
 
+// Does not rely on CXX11_STD_ATOMIC_128BIT
```
Is this comment a left-over or is it still true? If yes, please guard all of the code using 128Bit atomics (using `HPX_HAVE_CXX11_STD_ATOMIC_128BIT`).

hkaiser replied:
Sorry, I misread the comment, please disregard.

biddisco replied:
if the moody camel code used 128 bit atomics internally, would there be a simple way of checking (ie compiling with a certain compiler flag and triggering an error?)

cameron314 replied:
The `moodycamel::ConcurrentQueue` uses atomics of the `size_t` and `index_t` types from its traits, which are `std::size_t` by default. There's no use case for 128-bit atomics in the queue.

hkaiser replied:
@cameron314 thanks for clarifying!

msimberg commented on hpx/runtime/threads/policies/thread_queue_mc.hpp:
```diff
@@ -0,0 +1,1153 @@
+    // };
+    template <typename Mutex = std::mutex,
+        typename PendingQueuing = lockfree_lifo,
+        typename StagedQueuing = lockfree_lifo,
```
This template parameter needs to go (just like in the template for `shared_priority_queue_scheduler`)

msimberg replied:
What I mean here is that `shared_priority_queue_scheduler` passes its `TerminatedQueuing` template parameter as `StagedQueuing` to `thread_queue_mc`.

msimberg commented:
It looks like this is pretty much ready to go in. The only thing I don't particularly like is that most of `thread_queue_mc` just duplicates what's in `thread_queue`. Would inheriting from `thread_queue` be an option (assuming it doesn't add any overhead)?

biddisco commented:
I have another branch ready for a PR that builds upon this work. In the next PR, the shared_priority_queue is cleaned up (removing template params as above) and the thread_queue_mc is substantially redesigned. I put this thread_queue_mc in because it forms a nice starting point for the next layer of work. It's almost a copy/paste in this version but diverges significantly in the next version.
